okio/src/jvmMain/kotlin/okio/internal/ResourceFileSystem.kt

```diff
@@ -27,101 +27,77 @@ import okio.Sink
 import okio.Source
 import java.io.File
 import java.io.IOException
-import java.util.concurrent.ConcurrentHashMap
+import java.net.URL
 
 /**
  * A file system exposing Java classpath resources. It is equivalent to the files returned by
- * [ClassLoader.getResource].
+ * [ClassLoader.getResource] but supports extra features like [metadataOrNull] and [list].
  *
- * This file system does not implement merging of multiple paths from difference resources like
- * overlapping `.jar` files.
+ * If `.jar` files overlap, this returns an arbitrary element. For overlapping directories it unions
+ * their contents.
  *
  * ResourceFileSystem excludes `.class` files.
+ *
+ * This file system is read-only.
  */
 @ExperimentalFileSystem
 internal class ResourceFileSystem internal constructor(
-  private val classLoader: ClassLoader
+  classLoader: ClassLoader
 ) : FileSystem() {
-  private var jarCache = ConcurrentHashMap<Path, FileSystem>()
+  private var roots: List<Pair<FileSystem, Path>> = classLoader.toClasspathRoots()
 
   override fun canonicalize(path: Path): Path {
-    return "/".toPath() / path
+    return ROOT / path
   }
 
   override fun list(dir: Path): List<Path> {
-    val (fileSystem, fileSystemPath) = toSystemPath(dir) ?: return listOf()
-    return fileSystem.list(fileSystemPath)
+    val relativePath = dir.toRelativePath()
+    val result = mutableSetOf<Path>()
+    var foundAny = false
+    for ((fileSystem, base) in roots) {
+      try {
+        result += fileSystem.list(base / relativePath)
+          .map { it.removeBase(base) }
+        foundAny = true
+      } catch (_: IOException) {
```

Comment:
I wonder whether any non-FileNotFoundException here should be retained and later thrown in the suppression list?

Reply:
I think it might be confusing as it'll have the physical path, not the mapped path.

Reply:
Seems arguably confusing if a file permission issue or data corruption is silently swallowed and results in empty.  But happy to see real results and use any bugs raised as a trigger to address this.